File: Session_09_ShortestPaths/multi_language_implementations/dijkstra.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const V = 9
const INF = 999999

func minDistance(dist []int, sptSet []bool) int {
	min := INF
	min_index := -1

	for v := 0; v < V; v++ {
		if !sptSet[v] && dist[v] <= min {
			min = dist[v]
			min_index = v
		}
	}
	return min_index
}

func printSolution(dist []int) {
	fmt.Println("Vertex \t Distance from Source")
	for i := 0; i < V; i++ {
		if dist[i] == INF {
			fmt.Printf("%d \t INF\n", i)
		} else {
			fmt.Printf("%d \t %d\n", i, dist[i])
		}
	}
}

func runDijkstra(graph [V][V]int, src int) {
	var dist [V]int
	var sptSet [V]bool

	for i := 0; i < V; i++ {
		dist[i] = INF
		sptSet[i] = false
	}

	dist[src] = 0

	for count := 0; count < V-1; count++ {
		u := minDistance(dist[:], sptSet[:])
		sptSet[u] = true

		for v := 0; v < V; v++ {
			if !sptSet[v] && graph[u][v] != 0 && dist[u] != INF && dist[u]+graph[u][v] < dist[v] {
				dist[v] = dist[u] + graph[u][v]
			}
		}
	}

	printSolution(dist[:])
}

func main() {
	graph := [V][V]int{
		{0, 4, 0, 0, 0, 0, 0, 8, 0},
		{4, 0, 8, 0, 0, 0, 0, 11, 0},
		{0, 8, 0, 7, 0, 4, 0, 0, 2},
		{0, 0, 7, 0, 9, 14, 0, 0, 0},
		{0, 0, 0, 9, 0, 10, 0, 0, 0},
		{0, 0, 4, 14, 10, 0, 2, 0, 0},
		{0, 0, 0, 0, 0, 2, 0, 1, 6},
		{8, 11, 0, 0, 0, 0, 1, 0, 7},
		{0, 0, 2, 0, 0, 0, 6, 7, 0},
	}

	fmt.Println("=== Dijkstra's Single-Source Shortest Path ===")
	fmt.Println("Starting from Source Vertex: 0")
	fmt.Println()
	runDijkstra(graph, 0)

	fmt.Println()
	fmt.Print("Press Enter to exit...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}

File: Session_09_ShortestPaths/multi_language_implementations/dijkstra.rs
use std::io::{self, BufRead};

const V: usize = 9;
const INF: i32 = 999999;

fn min_distance(dist: &[i32; V], spt_set: &[bool; V]) -> usize {
    let mut min = INF;
    let mut min_index = 0;

    for v in 0..V {
        if !spt_set[v] && dist[v] <= min {
            min = dist[v];
            min_index = v;
        }
    }
    min_index
}

fn print_solution(dist: &[i32; V]) {
    println!("Vertex \t Distance from Source");
    for i in 0..V {
        if dist[i] == INF {
            println!("{} \t INF", i);
        } else {
            println!("{} \t {}", i, dist[i]);
        }
    }
}

fn run_dijkstra(graph: &[[i32; V]; V], src: usize) {
    let mut dist = [INF; V];
    let mut spt_set = [false; V];

    dist[src] = 0;

    for _ in 0..V - 1 {
        let u = min_distance(&dist, &spt_set);
        spt_set[u] = true;

        for v in 0..V {
            if !spt_set[v]
                && graph[u][v] != 0
                && dist[u] != INF
                && dist[u] + graph[u][v] < dist[v]
            {
                dist[v] = dist[u] + graph[u][v];
            }
        }
    }

    print_solution(&dist);
}

fn main() {
    let graph = [
        [0, 4, 0, 0, 0, 0, 0, 8, 0],
        [4, 0, 8, 0, 0, 0, 0, 11, 0],
        [0, 8, 0, 7, 0, 4, 0, 0, 2],
        [0, 0, 7, 0, 9, 14, 0, 0, 0],
        [0, 0, 0, 9, 0, 10, 0, 0, 0],
        [0, 0, 4, 14, 10, 0, 2, 0, 0],
        [0, 0, 0, 0, 0, 2, 0, 1, 6],
        [8, 11, 0, 0, 0, 0, 1, 0, 7],
        [0, 0, 2, 0, 0, 0, 6, 7, 0],
    ];

    println!("=== Dijkstra's Single-Source Shortest Path ===");
    println!("Starting from Source Vertex: 0\n");
    run_dijkstra(&graph, 0);

    println!();
    println!("Press Enter to exit...");
    let mut iterator = io::stdin().lock().lines();
    let _ = iterator.next();
}
